File: hw2/Shapes/src/Square.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.DoubleStream;

public class Square extends Rectangle implements Snappable {

    public Square(List<TwoDPoint> vertices) {
        super(vertices);
    }

    /**
     * Given a list of vertices assumed to be provided in a counterclockwise order in a two-dimensional plane, checks
     * whether or not they constitute a valid square.
     *
     * @param vertices the specified list of vertices in a counterclockwise order
     * @return <code>true</code> if the four vertices can form a square, <code>false</code> otherwise.
     */
    @Override
    public boolean isMember(List<? extends Point> vertices) {
        return vertices.size() == 4 &&
                DoubleStream.of(getSideLengths()).boxed().collect(Collectors.toSet()).size() == 1;
    }

    /**
     * Snaps the sides of the square such that each corner (x,y) is modified to be a corner (x',y') where x' is the
     * the integer value closest to x and y' is the integer value closest to y. This, of course, may change the shape
     * to a general quadrilateral, hence the return type. The only exception is when the square is positioned in a way
     * where this approximation will lead it to vanish into a single point. In that case, a call to {@link #snap()}
     * will not modify this square in any way.
     */
    @Override
    public Quadrilateral snap() {
    	List<TwoDPoint> gp = this.getPosition(); //original pos
    	List<TwoDPoint> gp2 = new ArrayList<TwoDPoint>(); // to become snapped pos
    	Quadrilateral q = new Quadrilateral(this.getPosition());
    	
    	for (int i = 0; i<=3 ; i++) {
    		gp2.add(new TwoDPoint(Math.round(gp.get(i).getX()),
    				Math.round(gp.get(i).getY())));
    	}
    	//if any of the points closed in on themselves
    	if(gp2.get(0).equals(gp2.get(1)) || gp2.get(0).equals(gp2.get(2)) || 
    			gp2.get(0).equals(gp2.get(3))|| gp2.get(1).equals(gp2.get(2))|| 
    			gp2.get(1).equals(gp2.get(3))|| gp2.get(2).equals(gp2.get(3))) {
    		return this;
    	}
    	q.setPosition(gp2);
    	return q;
    }
}

File: square.py
import math

from two_d_point import TwoDPoint
from rectangle import Rectangle
from quadrilateral import Quadrilateral


class Square(Rectangle):

    def __init__(self, *floats):
        super().__init__(*floats)
        if not self.__is_member():
            raise TypeError("A square cannot be formed by the given coordinates.")

    def __eq__(self, other):
        if isinstance(other, Square):
            x = 0
            for item in self.vertices:
                if item == other.vertices[x]:
                    x = x + 1
                else:
                    return False
        else:
            return False
        return True

    def __str__(self):
        strg = "Square: "
        for item in self.vertices:
            strg = strg + "(" + item.__str__() + "), "
        return strg

    def __is_member(self):
        sides = self.side_lengths()

        squ = (all(x == sides[0] for x in sides))
        return squ

    @staticmethod
    def __round(x):
        if x < 0:
            if x % 0.5 == 0:
                return math.floor(x)
            else:
                return round(x)
        if x > 0:
            if x % 0.5 == 0:
                return math.ceil(x)
            else:
                return round(x)
        return 0

    def snap(self):
        """Snaps the sides of the square such that each corner (x,y) is modified to be a corner (x',y') where x' is the
        integer value closest to x and y' is the integer value closest to y. This, of course, may change the shape to a
        general quadrilateral, hence the return type. The only exception is when the square is positioned in a way where
        this approximation will lead it to vanish into a single point. In that case, a call to snap() will not modify
        this square in any way."""

        rounded_vertices = [0, 0, 0, 0, 0, 0, 0, 0]
        x = 0
        for item in self.vertices:
            rounded_vertices[x] = Square.__round(item.x)
            x = x + 1
            rounded_vertices[x] = Square.__round(item.y)
            x = x + 1

        snapped_vertices = TwoDPoint.from_coordinates(rounded_vertices)

        bool1 = (all(a == snapped_vertices[0] for a in snapped_vertices))
        if bool1:
            return self

        quad = Quadrilateral(rounded_vertices)

        return quad
